tensorboard/components/experimental/plugin_util/plugin-host.ts

```diff
@@ -0,0 +1,100 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {IPC, MessageType, PayloadType, MessageCallback} from './message.js';
+
+const portIPCs = new Set<IPC>();
+const VERSION = 'experimental';
+const listeners = new Map<MessageType, MessageCallback>();
+
+// TODO(@psybuzz): replace this and the port cleanup logic in broadcast() with
+// a MutationObserver to notify us when iframes disconnect.
+const ipcToFrame = new Map<IPC, HTMLIFrameElement>();
+
+// The initial Window-level listener is needed to bootstrap only.
+// All further communication is done over MessagePorts.
+window.addEventListener('message', (event) => {
+  if (event.data !== `${VERSION}.bootstrap`) return;
+  const port = event.ports[0];
+  if (!port) return;
+  const frame = event.source ? event.source.frameElement : null;
+  if (!frame) return;
+  onBootstrap(port, frame as HTMLIFrameElement);
+});
+
+function onBootstrap(port: MessagePort, frame: HTMLIFrameElement) {
+  const portIPC = new IPC(port);
+  portIPCs.add(portIPC);
+  ipcToFrame.set(portIPC, frame);
+  port.start();
+
+  for (const [type, callback] of listeners) {
+    portIPC.listen(type, callback);
+  }
+}
+
+function _broadcast(
+  type: MessageType,
+  payload: PayloadType
+): Promise<PayloadType[]> {
+  for (const ipc of portIPCs) {
+    if (!ipcToFrame.get(ipc).isConnected) {
+      portIPCs.delete(ipc);
+      ipcToFrame.delete(ipc);
+    }
+  }
+
+  const promises = [...portIPCs].map((ipc) => ipc.sendMessage(type, payload));
+  return Promise.all(promises);
+}
+
+function _listen(type: MessageType, callback: MessageCallback) {
+  listeners.set(type, callback);
+  for (const ipc of portIPCs) {
+    ipc.listen(type, callback);
+  }
+}
+
+function _unlisten(type: MessageType) {
+  listeners.delete(type);
+  for (const ipc of portIPCs) {
+    ipc.unlisten(type);
+  }
+}
+
+export const broadcast = _broadcast;
+export const listen = _listen;
+export const unlisten = _unlisten;
+
+namespace tf_plugin {
```

Comment:
Namespaces are disallowed by go/tsstyle (publicly: https://github.com/google/gts/blob/83cb107010f7a956bb008291066304a9f26a0f34/tslint-rules.json#L43).  TensorBoard contains a fair number of them anyway, but let's not add more if possible.

Reply:
Good point.  @stephanwlee, could you weigh in on this?

There does seem to be alot of existing namespaces, ever since [this PR](https://github.com/tensorflow/tensorboard/pull/747/), but I think the Angular ng-tensorboard/ folder avoids this.

Reply:
1. plugin-host (at least until the tf_ng_tensorboard work is finished) will be loaded inside Polymer context which uses the namespace
2. (in anticipation of inevitable question) in ideal world, we want to share base code between host and guest impl except it is super hard to make ts_library build rule work with tf_web_library. The latter is used in mainly for polymer based TensorBoard. For now, since the experimental API is only used in context of Polymer anyways,  I decided to just deprecate the [es]module style and use namespace everywhere as v0.

Reply:
Thanks for the explanation.  It's unfortunate that Polymer code depends on these files.  I'll land as-is for now, but it seems we all agree that 'namespaces' are not ideal, and should be avoided if possible.